pybop/costs/_likelihoods.py

```diff
@@ -21,6 +21,31 @@ def __init__(self, problem: BaseProblem):
         self.n_data = problem.n_data
         self.minimising = False
 
+    def observed_fisher(self, y: dict, dy: np.ndarray) -> np.ndarray:
+        """
+        Compute the observed Fisher Information Matrix (FIM) for the given data.
+
+        The FIM is computed as the average of the squared gradients with respect to
+        the model parameters.
+
+        Parameters
+        ----------
+        y : dict
+            The model outputs for which to compute the FIM.
+        dy : np.ndarray
+            The sensitivities (gradients) of the model outputs with respect to parameters.
+
+        Returns
+        -------
+        np.ndarray
+            The observed Fisher Information Matrix.
+        """
+
+        # Calculate squared gradients element-wise
+        fisher_info = np.sum(np.square(dy), axis=0) / self.n_data
```

Comment:
After our discussion yesterday, I think it might be better to return the full observed FIM instead of the diagonal entries. Here's the implementation, thoughts?

Note, the gradient of the likelihood is what we are wanting here, which is returned from `self.__call__`

```suggestion
        _, grad = self.__call__(inputs, calc_grad = True)
        shaped_grad = grad.reshape(-1,1)
        fisher = (shaped_grad @ shaped_grad.T) / self.n_data
```